scrub_hub_vite/src/models/chat.ts

```diff
@@ -0,0 +1,197 @@
+import { fromBase64, toBase64 } from "../base64";
+import { ChatProps, MessageProps } from "../components/ChatComponent";
+import { ConversationListProps, ConversationProps } from "../components/ConversationListComponent";
+
+export type User = { name: string, id: number };
+function isUser(obj?: {[key: string]: unknown}): obj is User {
+	return !!(
+		obj &&
+		typeof obj.name === 'string' &&
+		typeof obj.id === 'number'
+	);
+}
+
+// Message level
+export class OutgoingMessage {
+	public constructor(private text: string, private id: number, private key: CryptoKey) { }
+
+	private async encryptMessage() {
+		const data = new TextEncoder().encode(this.text);
+		const iv = crypto.getRandomValues(new Uint8Array(12));
+		const encrypted = await crypto.subtle.encrypt({ name: 'AES-GCM', iv: iv }, this.key, data)
+		return {
+			'message': toBase64(encrypted),
+			'iv': toBase64(iv),
+		};
+	}
+
+	public async getData() {
+		return {
+			...await this.encryptMessage(),
+			id: this.id,
+		};
+	}
+}
+
+type IncomingMessageData = {
+	message: string,
+	username: string,
+	time: string, // must be convertable to Date
+	iv: string,
+}
+export function isIncomingMessageData(obj?: {[key: string]: unknown}): obj is IncomingMessageData {
+	if (!(
+		obj &&
+		typeof obj.message === 'string' &&
+		typeof obj.username === 'string' &&
+		typeof obj.time === 'string' &&
+		typeof obj.iv === 'string'
+	))
```

Comment:
again, what does all these checks serve?

Reply:
They are type guard functions, used to ensure the result of parsing JSON is what we expect.
Perhaps it would have been better to have the function checking the type to also perform the JSON parse. So more like `parseJsonAsIncomingMessageData(json: string): IncomingMessageData`

Reply:
Probably, it just seems a bit strange, the whole idea using typescript is literally to ensure things have a specific type?